iml-api/src/graphql.rs

```diff
@@ -397,6 +398,79 @@ impl QueryRoot {
         .await?;
         Ok(commands)
     }
+
+    /// Fetch the list of commands by ids, the returned
+    /// collection is guaranteed to match the input, if a command not found,
+    /// None is returned for that index.
+    #[graphql(arguments(
+        limit(description = "paging limit, defaults to 20"),
+        offset(description = "Offset into items, defaults to 0"),
+        ids(description = "The list of command ids to fetch, ids may be empty"),
+    ))]
+    async fn commands_by_ids(
+        context: &Context,
+        limit: Option<i32>,
+        offset: Option<i32>,
+        ids: Vec<i32>,
+    ) -> juniper::FieldResult<Vec<Option<Command>>> {
+        let ids: &[i32] = &ids[..];
+        let unordered_cmds: Vec<Command> = sqlx::query!(
+            r#"
+                SELECT
+                    c.id AS id,
+                    cancelled,
+                    complete,
+                    errored,
+                    created_at,
+                    array_agg(cj.job_id)::INT[] AS job_ids,
+                    message
+                FROM chroma_core_command c
+                JOIN chroma_core_command_jobs cj ON c.id = cj.command_id
+                WHERE (c.id = ANY ($3::INT[]))
+                GROUP BY c.id
+                OFFSET $1 LIMIT $2
+            "#,
+            offset.unwrap_or(0) as i64,
+            limit.unwrap_or(20) as i64,
+            ids,
+        )
+        .fetch_all(&context.pg_pool)
+        .map_ok(|xs: Vec<_>| {
+            xs.into_iter()
+                .map(|x| Command {
+                    id: x.id,
+                    cancelled: x.cancelled,
+                    complete: x.complete,
+                    errored: x.errored,
+                    created_at: x.created_at.format("%Y-%m-%dT%T%.6f").to_string(),
+                    jobs: {
+                        x.job_ids
+                            .unwrap_or_default()
+                            .into_iter()
+                            .map(|job_id: i32| {
+                                format!("/api/{}/{}/", Job::<()>::endpoint_name(), job_id)
+                            })
+                            .collect::<Vec<_>>()
+                    },
+                    logs: "".to_string(),
+                    message: x.message.clone(),
+                    resource_uri: format!("/api/{}/{}/", Command::endpoint_name(), x.id),
```

Comment:
Looks like a lot of duplication. Is it possible to extend the `commands` query to support optional id list?

Reply:
I tried, this `x` has very complicated synthetic type, derived by `sqlx` macro. Can be worked around with a transient local type though, say `TransientCommand`:
```
        let snapshots = sqlx::query_as!(TransientCommand, "SQL goes here")
              .fetch_all(&pool)
              .map(convertToCommand)
```

Reply:
Or `CommandTemp`